### ISSUE TITLE
New request: Appropedia, again

### ISSUE THREAD

**Issue by Sophivorus** [openzim/zim-requests#701]
- Website URL: https://www.appropedia.org/
- License: CC-BY-SA-4.0
- Desired ZIM Title: Appropedia
- Desired ZIM Description: Share knowledge to build rich, sustainable lives.
- Desired ZIM Icon: https://www.appropedia.org/w/images/8/84/Appropedia.png
- Language (ISO 639-3): en (actually, multilingual)
- Is this a MediaWiki?: yes

Hi everyone! I'd like to request a new ZIM file for Appropedia. I already requested this some time ago at [#371](https://github.com/openzim/zim-requests/issues/371), but encountered various issues, documented there as well as on [#1556](https://github.com/openzim/mwoffliner/issues/1556) and on an email conversation with Stephane and Emmanuel.

Looking back, the last problem was that my [Extension:CategoryLockdown](https://www.mediawiki.org/wiki/Extension:CategoryLockdown) was preventing read access to some (few) pages, and that was halting the script. The proposed solution was to generate a complete list of publicly available pages on Appropedia. I made a script that would output a very limited list of public pages and a ZIM file was generated successfully out of that. Today I finally expanded that script so that it outputs a list of ALL the public pages in (the main namespace of) Appropedia, see https://www.appropedia.org/scripts/generateKiwixList.php Note that it may take a few seconds to load. Also, redirects are excluded. Should I include them?

While delving into this, I installed Kiwix on my phone and downloaded Appropedia. The version available is from 2021-03-02, but I notice that https://farm.openzim.org/recipes/appropedia_en_all logs several successful runs 3 months ago. What does this mean? How come some recent runs were successful, if the problem with private pages was never solved? How come the most recent runs failed? And how come the latest version of Appropedia available on Kiwix is from 2021?

Anyway, thanks so much for developing and maintaining Kiwix. It's an awesome concept an software! More and more we're realizing that Appropedia's mission requires offline access, so we're eager to do our part to get a stable integration going!

PS: I improved CategoryLockdown so that it now outputs an error when trying to access a private page via the API, see for example https://www.appropedia.org/w/api.php?action=visualeditor&paction=parse&page=Dental_Procedures This new behavior was requested at some point in the conversation so that the scraper may one day detect private pages and react accordingly.

**Comment by Sophivorus** [comment]
Hi! It's been some months since this request. I'm afraid it's been forgotten, but we're still very much interested! Perhaps my message was a little too long? The bottom line is that I coded a script that outputs a CSV with all the pages that should be crawled, see https://www.appropedia.org/scripts/generateKiwixList.php (it may take a few seconds to load). Is this what you needed? Is there anything else we can do? Thanks!

**Comment by Popolechien** [comment]
Hola, sorry for the delay. Congrats on the revamp, this looks like a great resource and we should definitely make an effort to get it onboard.

I converted your .csv into a .tsv, ran the recipe and got the following error:

`[warn] [2024-03-07T14:46:31.167Z] Failed to get CSS from [https://www.appropedia.org/w/load.php?lang=en&modules=ext.Appropedia.styles|ext.CookieWarning.styles|ext.NoTitle,SearchThumbs|ext.ReadAloud.styles|ext.WikiVideos.styles|ext.smw.style|ext.smw.tooltip.styles|ext.srf.styles|ext.uls.pt|ext.visualEditor.desktopArticleTarget.noscript|mediawiki.page.gallery.styles|mediawiki.ui.button|mediawiki.widgets.styles|oojs-ui-core.icons,styles|oojs-ui.styles.icons-content,icons-editing-advanced,icons-editing-core,icons-editing-styling,icons-interactions,icons-layout,icons-media,icons-user,indicators|skins.poncho.styles&only=styles][warn] [2024-03-07T14:46:31.328Z] Failed to get CSS from [https://www.appropedia.org/wiki/?title=Mediawiki:offline.css&action=raw][error] [2024-03-07T14:46:34.216Z] Error downloading article Las_Malvinas_ecoladrillo_schoolroom_2012/Timeline[error] [2024-03-07T14:46:34.216Z] Failed to run mwoffliner after [185s]: {}`

([link to run log](https://farm.openzim.org/pipeline/0dd3458d-ef19-4128-b4a3-d3e1be9eee88/debug))

I have now put the recipe on hold (it was anyway regularly failing for almost a year), but figuring out next steps is a bit outside of my game. @kelson42 any suggestion?

**Comment by Sophivorus** [comment]
@Popolechien Hi, thanks for the reply and the attempt!

The first URL https://www.appropedia.org/w/load.php?lang=en&modules=ext.Appropedia.styles|ext.CookieWarning.styles|ext.NoTitle,SearchThumbs|ext.ReadAloud.styles|ext.WikiVideos.styles|ext.smw.style|ext.smw.tooltip.styles|ext.srf.styles|ext.uls.pt|ext.visualEditor.desktopArticleTarget.noscript|mediawiki.page.gallery.styles|mediawiki.ui.button|mediawiki.widgets.styles|oojs-ui-core.icons,styles|oojs-ui.styles.icons-content,icons-editing-advanced,icons-editing-core,icons-editing-styling,icons-interactions,icons-layout,icons-media,icons-user,indicators|skins.poncho.styles&only=styles seems to return CSS with no problems, not sure what could be wrong there.

The second URL https://www.appropedia.org/wiki/?title=Mediawiki:offline.css&action=raw is strange. I guess it's meant to get the CSS at MediaWiki:Offline.css, but instead it resolves to https://www.appropedia.org/?title=Mediawiki:offline.css&action=raw and gets the HTML of the main page. However, I tried the equivalent URL for the English Wikipedia https://en.wikipedia.org/wiki/?title=Mediawiki:offline.css&action=raw and it also returns the HTML of the main page, and in any case, neither Appropedia nor the English Wikipedia have a page titled MediaWiki:Offline.css

But those are warnings, not errors. The error seems to be with https://www.appropedia.org/Las_Malvinas_ecoladrillo_schoolroom_2012/Timeline However, that seems like a normal page, I can't find anything wrong or different about it. So I'm a bit flabbergasted too. What could be wrong?

**Comment by Popolechien** [comment]
I honestly don't know, @kelson42 hopefully will.

**Comment by Sophivorus** [comment]
Hi again! I just updated the script so that it outputs a TSV file with one title per line and no quotes around them. I think this is the preferred format? Also, I set "Content-Disposition: attachment", but not sure if that's desirable.

**Comment by benoit74** [comment]
Thank you for having fixed https://www.appropedia.org/scripts/generateKiwixList.php with removal of private pages.
I've restarted scraper and it went pretty far yesterday, but it failed on Israel_community_action which seems to be very slow to load (it got a 522).
This morning, https://www.appropedia.org/scripts/generateKiwixList.php is not responding anymore.

**Comment by Sophivorus** [comment]
*sigh* ..

Thank you for trying again! I see you did several runs, some with articleList=https://drive.farm.openzim.org/appropedia/generateKiwixList.tsv and others with articleList=https://www.appropedia.org/scripts/generateKiwixList.php

The first group I'm not sure why it failed. I reproduced the command on our server and got the same error. Perhaps it's because the TSV is relatively old? I guess it shouldn't matter, but the error log gives me no other clues and the pages on which the script failed seem perfectly normal. I even run mwoffliner for those pages only and the run completed ok.

As to the second group, I agree that generateKiwixList.php script was too slow and unreliable. So today I rewrote it as a maintenance script run from a cronjob. Now it's blazingly fast and creates a TSV file at https://appropedia.org/kiwix.tsv

So perhaps you can try another run with articleList=https://appropedia.org/kiwix.tsv ?

Thanks!

PS, I also created an updated 48x48 version of our logo at https://www.appropedia.org/logos/Appropedia-kiwix.png

**Comment by benoit74** [comment]
> The first group I'm not sure why it failed. 

The TSV had tabs at the end of each lines which are not supported (anymore?) by the scraper. I didn't realized fast enough it wasn't the proper URL ^^

> So perhaps you can try another run with the updated TSV

For now we have a new task which is doing great (TSV was back at some point, and I hurry to start the scraper at this moment ^^): https://farm.openzim.org/pipeline/adde019b-3b88-4274-b126-3fe0cfe423a0/debug ; finger crossed ^^

I have anyway updated the article list in the recipe for the new one (will not be used by current task obviously), thank you for the hard work

> I also created an updated 48x48 version of our logo 

Thanks, updated as well in the recipe (will not be used either by current task obviously)

**Comment by benoit74** [comment]
ZIMs have been produced succesfuly in DEV and they look good to me.

Only problem is that the home page is not the appropriate page but a tile of first pages of the TSV, and many are missing images which is ugly. This is a recipe configuration issue, I've already fixed the recipe and requested the ZIM again (in DEV, to check again everything works fine).

**Comment by benoit74** [comment]
We are blocked again by a very tiny bug, should be easy to fix next week: https://github.com/openzim/mwoffliner/issues/2163

**Comment by benoit74** [comment]
Issue solved, but we are now blocked by Cloudflare: https://github.com/openzim/mwoffliner/issues/2134

**Comment by benoit74** [comment]
@Sophivorus any chance you can whitelist few of our worker IPs in Cloudflare protections?

**Comment by benoit74** [comment]
New ZIM available at https://dev.library.kiwix.org/#lang=&q=appropedia

Unfortunately the home page is not that great, I did not realized that all icons are in fact links to search pages and not to articles. Inside the ZIM, all icons are hence missing the link since we do not support those search pages.

@Sophivorus is there any other article we could use for a better homepage inside the ZIM? Or should we craft one (we do this in quite a bunch of other ZIMs / Wikis, it also has the advantage that it is easier to remove some sections (e.g. the `Contact  Us`, `Leave us a message`, `Subscribe to the newsletter` and `Donate` buttons probably makes no sense when offline).

**Comment by Popolechien** [comment]
Great. Do you need both `nopic` and `maxi` versions, or should we only do the `maxi`?

**Comment by benoit74** [comment]
Since the problem behind homepage links is not going to be solved soon, may I propose to rename the custom homepage title from `Welcome_to_Appropedia/Kiwix` to `Welcome_to_Appropedia_Kiwix` or something similar? (the problem is linked to the slash in page title)

**Comment by Sophivorus** [comment]
Hi! Sorry for the delay, I just renamed `Welcome_to_Appropedia/Kiwix` to `Main_Page_for_Kiwix`. Regarding `nopic` and `maxi`, I think having only 'maxi' is ok!

**Comment by benoit74** [comment]
Unfortunately, this is still not working, links of the homepage are not clickable (not links at all in fact), I will have to investigate deeper.

**Comment by benoit74** [comment]
In fact it is working, only problem is that the link is only on the image in the gallery ... and in `nopic` we have no image, so the link is here, but content is empty ...

Is there any interest in having both the `nopic` (no pictures) and `maxi` (with pictures but without other medias like videos). Size of `maxi` is quite modest (510 MiB), `nopic` is significantly smaller (284 MiB). Or do we want to use another home page for `nopic` (totally feasible, no big impact for us from an infrastructure standpoint) ?

For now I've restarted the recipe with only `maxi` to at least confirm it works as expected.

**Comment by Sophivorus** [comment]
Having only `maxi` is fine !

**Comment by Sophivorus** [comment]
Hi again, I hope you're well!
I can see from https://farm.openzim.org/recipes/appropedia_en_all that the last run went ok!
However, when I open Kiwix and try to download Appropedia, I still see only the two old ZIMs from 2021.
Is something else needed to have them published? Or is it just a matter of time now?
Thanks!!

**Comment by kelson42** [comment]
@Sophivorus No, there is a problem

**Comment by Popolechien** [comment]
Recipe hadn't been moved to PROD. My bad. It's underway now.

**Comment by Sophivorus** [comment]
Thanks! I think deleting the old file would prevent confusions...

**Comment by benoit74** [comment]
`nopic` file marked for deletion from library/catalog, should be done in ~1h

**Comment by benoit74** [comment]
@Sophivorus recipe failed twice in a row due to 429 errors, looks like Cloudflare is blocking us, did you changed anything? Should we whitelist one of our workers IP?

**Comment by Sophivorus** [comment]
Hi! Yes, I had to up our defenses due to new abusive bots and it seems this caught your bot in the net. I could lower them again but I think whitelisting your IPs would be best. Could you tell me what they are? Thanks!

**Comment by benoit74** [comment]
List of IPs sent via mail since it is a bit sensitive information.

ZIM published in production: https://library.kiwix.org/#lang=eng&q=appropedia

Recipe reconfigured for automated quarterly updates.

**Comment by Sophivorus** [comment]
Hi! I just whitelisted the IPs, thanks! I also downloaded the published ZIM and it looks AMAZING! Thank you so much for the continued and kind support throughout these months. Hopefully, some of the issues tackled have improved Kiwix. Let me know if I can help you any other way!
